src/pages/LoginWithValidateCode2FAPage.js

```diff
@@ -58,21 +61,44 @@ class ValidateLogin2FANewWorkspacePage extends Component {
     }
 
     componentDidMount() {
-        // If the user has an active session already, they need to be redirected straight to the new workspace page
+        // If the user has an active session already, they need to be redirected straight to the relevant page
         if (this.props.session.authToken) {
             // In order to navigate to a modal, we first have to dismiss the current modal. But there is no current
-            // modal you say? I know, it confuses me too. Without dismissing the current modal, if they user cancels
-            // out of the new workspace modal, then they will be routed back to
-            // /v/<accountID>/<validateCode>/new-workspace and we don't want that. We want them to go back to `/` and
-            // by calling dismissModal(), the /v/... route is removed from history so the user will get taken to `/`
+            // modal you say? I know, it confuses me too. Without dismissing the current modal, if the user cancels out
+            // of the new workspace modal, then they will be routed back to
+            // /v/<accountID>/<validateCode>/workspace/123/card and we don't want that. We want them to go back to `/`
+            // and by calling dismissModal(), the /v/... route is removed from history so the user will get taken to `/`
             // if they cancel out of the new workspace modal.
             Navigation.dismissModal();
+            if (Permissions.canUseFreePlan(this.props.betas)) {
+                this.rerouteToRelevantPage();
+            }
+        }
+    }
 
-            if (_.isEmpty(this.props.betas)) {
-                setRedirectToWorkspaceNewAfterSignIn(true);
-            } else {
+    componentDidUpdate() {
+        // Betas can be loaded a little after a user is authenticated, so check again if the betas have been updated
+        if (this.props.session.authToken && Permissions.canUseFreePlan(this.props.betas)) {
```

Comment:
Should this also check that `this.props.betas` exists before calling `canUseFreePlan()`?

Reply:
It will always exist because of the default props of this component having it set to `[]`. We could check if it's empty, but `Permissions.canUseFreePlan` doesn't need the argument to be a non-empty array so I feel like that check might be redundant.

Reply:
Here is a better description of why I thought this could have been a problem:

1. Component mounts
2. Component is updated with a session auth token, but betas haven't loaded yet
3. Permission check happens with an empty beta list
4. Betas finish loading
5. Now the beta check from 3 was stale and wrong because of the race condition

It's possible this isn't actually a problem because the first permission check would always return `false` if the beta array is empty, and there would be no redirect. Then, once the betas finish loading, there would be another component update with the proper beta data and the permission check would run and return `true` and then the redirect happens.

I think we are OK? It's quite the "gotcha" though and relies on the first permission check returning false with an empty beta array. If someone were to ever implement a permission check that did opposite logic (like `cannotUseFreePlan`, then the race condition would happen).

Reply:
I think we are ok!  And ahh yeah I haven't seen the `cannotUse` type of beta checks, so I didn't consider them. Your reasoning around the beta checks is right and the component loads endlessly Until we CAN prove the user is in the beta. Hypothetically a user who isn't in any betas who landed on this page would see an endless loader, but such a user shouldn't be able to access the link that brings them here in the first place.